### ISSUE TITLE
Switch Off UE BOOM 2 failed

### ISSUE THREAD

**Issue by freMea** [Shingyx/BoomSwitch#5]
App version 1.3
Android 9
Speaker firmware is last: 2.4.7
All actions works with official app.

Switch On works.

Once the speaker is connected on my phone, I try to switch Off. It fails returning : Connection failed.

**Comment by Shingyx** [comment]
Thanks for reporting this issue!

Which exact message are you seeing? Is it the message "Connection failed. Is \<speaker name> in range"? If so, it implies that either the speaker's BLE (Bluetooth Low Energy) server is not running, or that the phone is failing to connect to it.

Note that the official app uses BLE to power the speaker on, but some other communication method using standard Bluetooth to power it off. The BOOM Switch app uses BLE to both power on and power off. This could be related to the issue you're seeing.

Can you please confirm again if your speaker is on the latest firmware using the official app? Also, can you please confirm the exact message you're seeing? 

Thanks.

**Comment by freMea** [comment]
Error message is : "Connection failed. Is <speaker name> in range?"

I confirm the firmware is the last.

**Comment by freMea** [comment]
It seems it works in some cases. When the speaker is on but not in use for some time, the switch off command from your app appears to connect to speaker.

**Comment by Shingyx** [comment]
Interesting how it works in some cases. Does it seem random when it works and when it doesn't, or is there a pattern? This is the first time I've heard of an issue like this and I can't reproduce it myself. I have a _slight_ feeling this issue could be related to the phone that's being used, or maybe the speaker's hardware.

I have seen some weird behaviour when BOOM Switch is used while the speaker is connected to a different device, but that doesn't sound like what you're experiencing.

Regarding the behaviour of how it connects to the speaker instead of turning it off, that is actually by design, implemented to fix the issue https://github.com/Shingyx/BoomSwitch/issues/1.

**Comment by freMea** [comment]
For some reason, the switch OFF command have worked during my last tests after reading your message. No matter how many time the speaker is ON and connected and if is silent or playing audio. Weird because, of course, I previously tried multiple things to fix the issue, including reboot my phone, and nothing has changed. But then as soon as I open the _Ultimate Ears_ official app, the issue came back. But this last action doesn’t always cause the issue. I really don’t know how to trigger the issue for sure.

However I found a fix when it happens. I think it always work but not sure. I kill the BoomSwitch app and clear its cache and data. Then command to switch off works again. 

In Tasker I tried to add a task to automate cache+data erasing but failed to clear all the data from the app. System reports there is 28.67Kb left. Here are the data directories I delete (I’m rooted) where %par1 = com.github.shingyx.boomswitch to no avail :
* /data/data/%par1/shared_prefs
* /storage/emulated/0/Android/data/%par1 (doesn’t exist)
* /data/user_de/%par1/shared_prefs (doesn’t exist)

Do you know what other files or directories I must delete to get the same result as the “clear data” button ?

My phone is a OnePlus 6, OSS 9.0.9 (last firmware from PIE).

**Comment by Shingyx** [comment]
I tried reproducing this with my Galaxy S10 running Android 10 and my personal BOOM 3 speaker, but unfortunately I can't reproduce this issue. When the speaker is on, I can turn it off using BOOM Switch with no issues while the official app is still running in the background. I've tested this multiple times, turning the speaker on using the physical power button, the official app, and BOOM Switch.

Earlier I mentioned this issue could happen because of two possibilities - either the speaker's BLE (Bluetooth Low Energy) server stops running, or the phone is failing to connect to it.

I _don't think_ the BOOM 2 and BOOM 3 are much different to each other regarding this feature, so I feel it's more likely that the phone is failing to connect to the speaker's BLE server. It's possible that Android 10 introduces improvements around Bluetooth that prevents this issue from happening. The fact that clearing BOOM Switch's cache+data fixes the issue further suggests that Android 10 may be the solution.

Sorry I couldn't be much help, but I can't do much more without acquiring more Android devices and/or speakers for testing this. My best advice to attempt an upgrade to Android 10, or avoid using the official app if possible.
